2/contrib/s2i/run

```diff
@@ -11,11 +11,117 @@ source /usr/local/bin/kube-slave-common.sh
+IMAGE_CONFIG_DIR="/opt/openshift/configuration"
+
+# temp directory where template expansion and configs will be copied to before
+# being copied to the $JENKINS_HOME
+IMAGE_CONFIG_GEN_DIR=$(mktemp --dry-run -d -p "$JENKINS_HOME")
+
+# file that holds the timestamp/signature of config files mounted through configmap
+CONFIG_TIMESTAMP="${CONFIG_TIMESTAMP:-config-timestamp}"
+
+log() {
+  local level=$1; shift
+  local datetime=$(date "+%b %d, %Y %r")
+  echo "${datetime} ${level}: $*"
+}
+
+log_info() {
+  log INFO "$@"
+}
+
+log_warn() {
+  log WARN "$@"
+}
+
+config_changed() {
+  # don't worry about configurations if there isn't a dir
+  # but note that the image we are building already has the dir
+  [ -d ${IMAGE_CONFIG_DIR} ] || return 1
+
+  local img_config_timestamp="$IMAGE_CONFIG_DIR/$CONFIG_TIMESTAMP"
+
+  # treat absense of timestamp in configmap as timestamp changed to be on the
+  # safe side so that configuration files are generated
+  [ -f "$img_config_timestamp" ] || return 0
+
+  local our_timestamp="$JENKINS_HOME/$CONFIG_TIMESTAMP"
+  touch -a "$our_timestamp"
+
+  # return 0 if files are different
+  [[ $(cmp -s "$img_config_timestamp" "$our_timestamp") != 0 ]]
+}
+
+generate_jenkins_config()  {
+  local config_file="config.xml"
+  local config_tpl_path="$IMAGE_CONFIG_DIR/$config_file.tpl"
+
+  [ -f "$config_tpl_path" ] || {
+    log_warn "Skipping generation of '$config_file' as $config_tpl_path is not found"
+    return 0
+  }
+
+  # If it contains a template (tpl) file, we can do additional manipulations
+  # to customize the configuration.
+  export KUBERNETES_CONFIG=$(generate_kubernetes_config)
+
+  for name in $(get_is_names); do
+    log_info "... adding image ${name}:latest as Kubernetes slave"
+  done
+  log_info "Generating kubernetes-plugin configuration (${config_tpl_path})"
+  envsubst < "$config_tpl_path" > "${IMAGE_CONFIG_GEN_DIR}/$config_file"
+}
+
+generate_credentials()  {
+  local cred_file="credentials.xml"
+  local cred_tpl_path="$IMAGE_CONFIG_DIR/$cred_file.tpl"
+
+  [ -f "$cred_tpl_path" ] || {
+    log_warn "Skipping generation of '$cred_file' as $config_tpl_path is not found"
+    return 0
+  }
+
+  [ -z "${KUBERNETES_CONFIG}" ] && {
+    log_warn "Skipping generaton of kubernetes-plugin credentials as KUBERNETES_CREDENTIALS is blank"
+    return 0
+  }
+
+  log_info "Generating kubernetes-plugin credentials ($cred_file) ..."
+  export KUBERNETES_CREDENTIALS=$(generate_kubernetes_credentials)
+
+  # Fix the envsubst trying to substitute the $Hash inside credentials.xml
+  export Hash="\$Hash"
+  envsubst < "$cred_tpl_path" > "${IMAGE_CONFIG_GEN_DIR}/$cred_file"
+}
+
+copy_config_files() {
+  log_info "Copying Jenkins configuration to $JENKINS_HOME"
+  cp -frL "$IMAGE_CONFIG_DIR/"* "$JENKINS_HOME/"
+  cp -frL "$IMAGE_CONFIG_GEN_DIR/"* "$JENKINS_HOME/"
+
+  # we don't need the templates
+  log_info "Removing template files (*.tpl) from $JENKINS_HOME"
+  rm -f "$JENKINS_HOME/"*.tpl
```

Comment:
Is this remove really necessary? My concerns is that we face the same problem we had some time ago.

Reply:
no it shoudn't cause any errors as $JENKINS_HOME is rw and not read-only. If JENKINS_HOME is `ro ` then that is a much bigger problem as nothing can write to it.

Reply:
It is not a problem on an error but in performance problems we faced in the past . cc @kbsingh what do you think?

Reply:
Wasn't the root cause for performance issue that we were copying 400MB of plugin which is like close to 300 files? 

A reason to do this way using a tmp dir is to ensure that failure of an in between step do not leave the system in a bad state.

Reply:
Ok I think we can remove safely because exploring the issue I am mention (https://github.com/openshiftio/openshift.io/issues/2612) I see that it was because of a move not a remove. So go ahead.